File: CivitasJava/src/civitas/MazoSorpresas.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas;

import java.util.ArrayList;
import java.util.Collections;

/**
 *
 * @author luis
 */
public class MazoSorpresas {
    private ArrayList<Sorpresa> sorpresas;
    private boolean barajada;
    private int usadas;
    private boolean debug;
    private ArrayList<Sorpresa> cartasEspeciales;
    private Sorpresa ultimaSorpresa;
    
    private void init(){
        sorpresas = new ArrayList<Sorpresa>();
        cartasEspeciales = new ArrayList<Sorpresa>();
        barajada = false;
        usadas = 0;
    }
    
    MazoSorpresas(boolean debug){
        this.debug = debug;
        init();
        
        if(debug){
            Diario diary = Diario.getInstance();
            diary.ocurreEvento("debug activado");
        }
    }
    
    MazoSorpresas(){
        init();
        debug = false;
    }
    
    void alMazo(Sorpresa s){
        if(!barajada)
            this.sorpresas.add(s);
    }
    
    Sorpresa siguiente(){
        if(!barajada || this.sorpresas.size() == usadas && !debug)
        {
            barajada = true;
            Collections.shuffle(sorpresas);
            usadas = 0;
        }
        
        usadas++;
        this.ultimaSorpresa = this.sorpresas.get(0);
        this.sorpresas.remove(0);
        this.sorpresas.add(ultimaSorpresa);
        
        return ultimaSorpresa;
    }
    
    Sorpresa getUltimaSorpresa(){
        return ultimaSorpresa;
    }
    
    void inhabilitarCartaEspecial(Sorpresa sorpresa){
        if(sorpresas.contains(sorpresa)){
            this.sorpresas.remove(sorpresa);
            this.cartasEspeciales.add(sorpresa);
            Diario diary = Diario.getInstance();
            diary.ocurreEvento("Sorpresa añadida a cartas especiales");
        }
    }
    
    void habilitarCartaEspecial(Sorpresa sorpresa){
         if(cartasEspeciales.contains(sorpresa)){
            this.cartasEspeciales.remove(sorpresa);
            this.sorpresas.add(sorpresa);
            Diario diary = Diario.getInstance();
            diary.ocurreEvento("Sorpresa añadida a mazo sorpresas");
        }
    }
}

File: CivitasRuby/pkg/CivitasRuby-0.0.1/lib/mazo_sorpresas.rb
# encoding:utf-8
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

require_relative 'sorpresa.rb'

module Civitas
  
  class MazoSorpresas

    def initialize()
      @sorpresas = Array.new
      @cartasEspeciales = Array.new
      @barajada = false
      @usadas = 0
      @debug = false
    end
    
    def self.new_mazo_sorpresas1(debug)
      new
      @debug = debug
      
      if(debug)
        diary = Diario.instance
        diary.ocurre_evento("debug activado")
      end
      
    end
    
    def self.new_mazo_sorpresas2()
        new
    end
   
    
    def al_mazo(s)
      if !@barajada
        @sorpresas << s
      end
    end
    
    def siguiente()
      if !@barajada && @sorpresas.length == @usadas && !@debug
        @barajada = true
        @sorpresas.shuffle
        @usadas = 0
      end
      @usadas = @usadas + 1
      @ultimaSorpresa = @sorpresas[0]
      @sorpresas.delete_at(0)
      @sorpresas << @ultimaSorpresa
      
      return @ultimaSorpresa
    end
    
    def inhabilitar_carta_especial(sorpresa)
      if @sorpresas.include?(sorpresa)
        @sorpresas.delete(sorpresa)
        @cartasEspeciales << sorpresa
        diary = Diario.instance
        diary.ocurre_evento("Sorpresa añadida a cartas especiales")
      end
    end
    
    def habilitar_carta_especial(sorpresa)
      if @cartasEspeciales.include?(sorpresa)
        @cartasEspeciales.delete(sorpresa)
        @sorpresas << sorpresa
        diary = Diario.instance
        diary.ocurre_evento("Sorpresa añadida a mazo sorpresas")
      end
    end
    
    private_class_method :new
  end
  
end
